File: LeetCode/Python Solutions/Distribute Coins in Binary Tree Solved/distribute.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def distributeCoins(self, root: Optional[TreeNode]) -> int:
        self.moves = 0
        
        def dfs(node):
            if not node:
                return 0
            
            left_excess = dfs(node.left)
            right_excess = dfs(node.right)
            
            # Current node's total excess coins
            excess = node.val + left_excess + right_excess - 1
            
            # Total moves is the sum of the absolute values of excess coins from children
            self.moves += abs(left_excess) + abs(right_excess)
            
            return excess
        
        dfs(root)
        return self.moves

File: LeetCode/Java Solutions/Distribute Coins in Binary Tree Solved/distribute.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int moves;
    
    public int distributeCoins(TreeNode root) {
        moves = 0;
        dfs(root);
        return moves;
    }
    
    private int dfs(TreeNode node) {
        if (node == null) {
            return 0;
        }
        
        int leftExcess = dfs(node.left);
        int rightExcess = dfs(node.right);
        
        // Current node's total excess coins
        int excess = node.val + leftExcess + rightExcess - 1;
        
        // Total moves is the sum of the absolute values of excess coins from children
        moves += Math.abs(leftExcess) + Math.abs(rightExcess);
        
        return excess;
    }
}
